src/python/pants/backend/tools/semgrep/subsystem.py

```diff
@@ -0,0 +1,121 @@
+# Copyright 2023 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import Iterable
+
+from pants.backend.python.goals import lockfile
+from pants.backend.python.goals.export import ExportPythonTool, ExportPythonToolSentinel
+from pants.backend.python.goals.lockfile import GeneratePythonLockfile
+from pants.backend.python.subsystems.python_tool_base import (
+    ExportToolOption,
+    LockfileRules,
+    PythonToolBase,
+)
+from pants.backend.python.target_types import ConsoleScript
+from pants.backend.python.util_rules.pex_requirements import GeneratePythonToolLockfileSentinel
+from pants.core.goals.generate_lockfiles import GenerateToolLockfileSentinel
+from pants.engine.rules import Rule, collect_rules, rule
+from pants.engine.target import Dependencies, FieldSet, SingleSourceField, Target
+from pants.engine.unions import UnionRule
+from pants.option.option_types import ArgsListOption, BoolOption, SkipOption
+from pants.util.strutil import softwrap
+
+
+@dataclass(frozen=True)
+class SemgrepFieldSet(FieldSet):
+    required_fields = (SingleSourceField, Dependencies)
```

Comment:
Oh this is interesting. I have long standing thoughts on how a hybrid plugin would work with targets.

Anything with a source seems like a good way of doing that

🙌

Reply:
Cool 👍 

(Coincidentally, your comment makes me realise that the `Dependencies` field isn't used any more, so I should get rid of it. Note to self.)

Reply:
Hmm I feel like there's an example of targets that operate on a level that isn't single-file. Maybe this should be any source field? 

You can still run it on individual files, but you'll match more targets.

Reply:
were you still getting rid of the dependencies field?